### ISSUE TITLE
Grafana Table doesn't merge queries results with Victoria Metrics datasource

### ISSUE THREAD

**Issue by Keiske** [VictoriaMetrics/VictoriaMetrics#720]
**Describe the bug**
Switching datasource from Prometheus to Victoria Metrics in Grafana Tables breaks row merging for the same data. We are trying to use VictoriaMetrics as long-term historical metrics storage for our Kubernetes in-cluster Prometheus. Have configured Prometheus to send data to VictoriaMetrics via remote_write api. 

**To Reproduce**
1. Install VictoriaMetrics single node helm chart.
2. Add remote_write to VM in Prometheus config.
3. Change datasource from prometheus to VM in grafana dashboards with tables with multiple queries. No any other settings in Grafana changed.
4. Queries results doesn't merged anymore. See screenshots.
5. Change datasouce back to Prometheus - tables look as it should.

**Expected behavior**
Multiple queries results with same lebels should be merged to rows in Grafana Tables panel.

**Screenshots**
Prometheus datasource table result:
![prom_grafana_merge2](https://user-images.githubusercontent.com/1940791/91213801-d06a1300-e71a-11ea-9313-3fd02299f25e.JPG)

VictoriaMetrics datasource table result with same queries and settings:
![vm_grafana_merge1](https://user-images.githubusercontent.com/1940791/91213787-c9db9b80-e71a-11ea-94ae-a22cd251eadd.JPG)

**Version**
The line returned when passing `--version` command line flag to binary. For example:
```
$ ./victoria-metrics-prod --version
victoria-metrics-20200815-125320-tags-v1.40.0-0-ged00eb3f3
```

**Comment by hagen1778** [comment]
Hi @Keiske! Could you pls check the actual response in both cases and compare? Would be nice to post it here if possible.

**Comment by Keiske** [comment]
> Hi @Keiske! Could you pls check the actual response in both cases and compare? Would be nice to post it here if possible.

Sure. Here it is. Looks very similar, but with switched order of metrics in results. 


First value column query result:
Prometheus:
`{"status":"success","data":{"resultType":"vector","result":[{"metric":{"pod":"ag-postgres-portal-0"},"value":[1598532794,"0.25"]},{"metric":{"pod":"ag-postgres-user-0"},"value":[1598532794,"0.25"]}]}}`

VictoriaMetrics:
`{"status":"success","data":{"resultType":"vector","result":[{"metric":{"pod":"ag-postgres-user-0"},"value":[1598532794,"0.25"]},{"metric":{"pod":"ag-postgres-portal-0"},"value":[1598532794,"0.25"]}]}}`


Second value column query results:
Prometheus:
`{"status":"success","data":{"resultType":"vector","result":[{"metric":{"pod":"ag-postgres-portal-0"},"value":[1598532794,"0.0908893273381095"]},{"metric":{"pod":"ag-postgres-user-0"},"value":[1598532794,"0.0875724036524064"]}]}}`

VictoriaMetrics:
`{"status":"success","data":{"resultType":"vector","result":[{"metric":{"pod":"ag-postgres-user-0"},"value":[1598532794,"0.08627032711112406"]},{"metric":{"pod":"ag-postgres-portal-0"},"value":[1598532794,"0.09153162373333341"]}]}}`

![vmag](https://user-images.githubusercontent.com/1940791/91449325-a2eca900-e883-11ea-835f-ac3338f921cf.JPG)

**Comment by hagen1778** [comment]
Thanks! Weird, content looks identical except order. But order is consistent for both VM requests so it shouldn't be a case. Have you tried to build any other tables based on VM datasource? I wonder if it is Grafana bug...

**Comment by valyala** [comment]
@Keiske , try wrapping queries into `sort()` function. This should guarantee consistent order of the returned metrics.

**Comment by balabalazhoucj** [comment]
Hi!I have same question.My environment uses v1.40.0-cluster.

![WX20200909-105454](https://user-images.githubusercontent.com/10136473/92549041-01a22180-f28b-11ea-9330-36a693237c7c.png)

But I change grafana time range is ok(from now-5m to now-1m),my prometheus scrape_interval is 1m.

![WX20200909-105416](https://user-images.githubusercontent.com/10136473/92549052-06ff6c00-f28b-11ea-895a-bd084c89f867.png)

```
/ # ./vminsert-prod -version
vminsert-20200815-132700-tags-v1.40.0-cluster-0-gd9f7ea1c6
```

When i use v1.37.2-cluster is ok

**Comment by hagen1778** [comment]
Hi @balabalazhoucj! Have you tried to set `instant` button to your queries like in the screenshot from @Keiske?

**Comment by Muxa1L** [comment]
@hagen1778 I think its because when "instant" is used it returns time for as <current time> - 30s (or current value of search.latencyOffset param set for vmselect) instead of actual time for this metric. And because queries may complete in different time this timestamps may differ for 1-2-10-inf ms, and it breaks the table. 
Examples. 
Not OK: (3ms offset for metrics scraped in one time)
![image](https://user-images.githubusercontent.com/9287490/93459297-2f6e1280-f8ea-11ea-82af-2a076980d902.png)
OK (after 10 refreshes):
![image](https://user-images.githubusercontent.com/9287490/93460194-87f1df80-f8eb-11ea-905f-02986efd65c8.png)
Nothing was changed in queries, and dashboards, only few refreshes.

**Comment by Keiske** [comment]
@valyala Sorting results didn't help. 

@hagen1778 Yes, as @Muxa1L said, our issue looks just like this. Refreshing dashboard in browser sometimes makes table marge correct. About 1 of 10 page refreshes. But how can we fix it to always merge results in table correctly, just like Prometheus datasource does.

**Comment by Muxa1L** [comment]
By the way, when you disable "instant" thus running ranged query - last results of query are correct.

**Comment by hagen1778** [comment]
@Muxa1L Good catch! Instant queries are sent to `/query` handler and response contains timestamps with milliseconds precision. And regular queries are sent to `/query_range` and response contains timestamps with second precision. Should VM round timestamps up to seconds for instant queries @valyala ?

**Comment by Muxa1L** [comment]
@hagen1778 I think it will be better to return the timestamp of last metric value
Example single metric and results for instant query (to path /query):
{"result":[{"metric":{"__name__":"go_cpu_count","instance":"self","job":"victoria-metrics"},"value":[1600421855.726,"3"]}]}
1600421855.726 is approx 2020-09-18 12:37:35 - is the <time when query was executed> - 30s
, and ranged (for last minute)
{"result":[{"metric":{"__name__":"go_cpu_count","instance":"self","job":"victoria-metrics"},"values":[[1600421820,"3"],[1600421835,"3"],[1600421850,"3"],[1600421865,"3"],[1600421880,"3"]]}]}
1600421880 is 2020-09-18 12:38:00, and is a correct time of last scrape

**Comment by starsliao** [comment]
@hagen1778 
I also have this problem. I found that there will be such a problem since v1.39.0, but it is normal in v1.38.1. Because the grafana table cannot be used, I can only use the v1.38.1 version now. I hope this problem can be solved.

- v1.39.0
![x](https://user-images.githubusercontent.com/3349611/93727713-f5a15280-fbee-11ea-9cab-883ebf158f01.jpg)


- v1.38.1
![image](https://user-images.githubusercontent.com/3349611/93727741-27b2b480-fbef-11ea-880a-a86d7de6bcf7.png)

**Comment by Muxa1L** [comment]
@hagen1778 Nvm my previous comment. Returning timestamps with seconds precision will be enough. Prometheus also returns with seconds precision. And victoriaMetrics returned timestamps with seconds precision before v1.39.0, as @starsliao noticed

**Comment by valyala** [comment]
The issue must be fixed in the following commits:
* Single-node VictoriaMetrics - 2eb72e09ab7bd3208cfbe08a08938f043ea4e876
* Cluster VictoriaMetrics - 07c622633476d61e9ceef646a41e9e3f06fd7fd4

The bugfix rounds default `time` value to seconds when the query to [/api/v1/query](https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries) doesn't contain `time` query arg. This is a workaround, which reduces the probability of the original issue. The proper fix should be applied on Grafana side - it must pass `time` query arg with each query to `/api/v1/query`.

**Comment by Muxa1L** [comment]
@valyala grafana passes time to queries. But it seems that it is not counted anywhere.
![изображение](https://user-images.githubusercontent.com/9287490/93860476-d4f8fb80-fcc7-11ea-879c-71f7181097f6.png)

**Comment by valyala** [comment]
@Muxa1L , thanks for the spot! It must be fixed in the following commits:

* Single-node VictoriaMetrics: 3ba507000ca78b35cbe2db8106131c8376450ba6
* Cluster VictoriaMetrics: 1fce79518a58956c8cc8d62654907edbefe53a04

Unfortunately these commits weren't included in [v1.41.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases), but they will be included in the next release.

**Comment by Muxa1L** [comment]
@valyala,  great! Now it works correctly, thanks for fixes!

**Comment by boazjohn** [comment]
Is it advised to use rollback to v1.38.1? Is this fix going to to be released soon? 
Using 1.41.1 and facing the issue.

**Comment by valyala** [comment]
> Is it advised to use rollback to v1.38.1?

Unfortunately it is impossible to downgrade from v1.41.* to older releases due to on-disk data format change. See [release notes for v1.41.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.41.0) for details. So it is better waiting for the next release or building VictoriaMetrics from sources according to the following docs:

* [How to build single-node version of VictoriaMetrics](https://victoriametrics.github.io/#how-to-build-from-sources)
* [How to build cluster version of VictoriaMetrics](https://victoriametrics.github.io/Cluster-VictoriaMetrics.html#building-from-sources)

> Is this fix going to to be released soon?

The fix will be included in the upcoming release, which is going to be published in the next couple of days.

**Comment by valyala** [comment]
The bugfix is available starting from [v1.42.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing the bug as fixed.
